pkg/security/podsecuritypolicy/provider.go

```diff
@@ -229,10 +229,28 @@ func (s *simpleProvider) ValidatePodSecurityContext(pod *api.Pod, fldPath *field
 			}
 
 			if fsType == extensions.HostPath {
-				if !psputil.AllowsHostVolumePath(s.psp, v.HostPath.Path) {
+				if len(s.psp.Spec.AllowedHostPaths) == 0 {
+					// If no allowed paths are specified then allow any path
+					// since host path volumes are allowed.
+					continue
```

Comment:
continuing in "success" cases assumes we'll never add other things that need checking later on in the loop. keep this structured to `continue` only to short-circuit the current volume checking in case of an error

Reply:
@liggitt this is keeping the same logic that we had in the util method before. I was trying to decrease the number of changes required to make this change.  You'd prefer i set a variable then below check it before emitting every error?

Reply:
I think what Liggitt is saying is that you're now `continue`ing over stuff like flex volume validation further on. I'm personally confused why `len(s.psp.Spec.AllowedHostPaths) == 0` would cause you to skip the flex volume checks. 

Maybe just change the above if statement to:

```go
if fsType == extensions.HostPath && len(s.psp.Spec.AllowedHostPaths) > 1 {
    // ...
}
```

Reply:
@ericchiang It would never skip flex volume checks if the fs type was flex volume because a volume can't be two types at the same time. So if it reached the existing allowed host path check then it must be a host volume and nothing else.

Reply:
> You'd prefer i set a variable then below check it before emitting every error?

No, I'd prefer to only `continue` early in cases where we added an error. The current change adds `continue` statements in non-error cases.

> It would never skip flex volume checks if the fs type was flex volume because a volume can't be two types at the same time. So if it reached the existing allowed host path check then it must be a host volume and nothing else.

Continuing early in non-error cases assumes all other checks will be specific to other volume types, which is an assumption I'm trying to avoid adding here.

> I was trying to decrease the number of changes required to make this change.

I do agree on minimal change... I think it can be contained entirely to this block and the AllowsHostVolumePath helper:

```diff
--- a/pkg/security/podsecuritypolicy/provider.go
+++ b/pkg/security/podsecuritypolicy/provider.go
@@ -229,10 +229,27 @@ func (s *simpleProvider) ValidatePodSecurityContext(pod *api.Pod, fldPath *field
                        }

                        if fsType == extensions.HostPath {
-                               if !psputil.AllowsHostVolumePath(s.psp, v.HostPath.Path) {
+                               pathIsAllowed, mustBeReadOnly := psputil.AllowsHostVolumePath(s.psp, v.HostPath.Path)
+                               if !pathIsAllowed {
                                        allErrs = append(allErrs, field.Invalid(
                                                field.NewPath("spec", "volumes").Index(i).Child("hostPath", "pathPrefix"), v.HostPath.Path,
                                                fmt.Sprintf("is not allowed to be used")))
+                               } else if mustBeReadOnly {
+                                       // Ensure all the VolumeMounts that use this volume are readonly
+                                       for i, c := range pod.Spec.InitContainers {
+                                               for j, cv := range c.VolumeMounts {
+                                                       if cv.Name == v.Name && !cv.ReadOnly {
+                                                               allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "initContainers").Index(i).Child("volumeMounts").Index(j).Child("readOnly"), cv.ReadOnly, "must be read-only"))
+                                                       }
+                                               }
+                                       }
+                                       for i, c := range pod.Spec.Containers {
+                                               for j, cv := range c.VolumeMounts {
+                                                       if cv.Name == v.Name && !cv.ReadOnly {
+                                                               allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "containers").Index(i).Child("volumeMounts").Index(j).Child("readOnly"), cv.ReadOnly, "must be read-only"))
+                                                       }
+                                               }
+                                       }
                                }
                        }
```

```diff
--- a/pkg/security/podsecuritypolicy/util/util.go
+++ b/pkg/security/podsecuritypolicy/util/util.go
@@ -173,25 +173,29 @@ func GroupFallsInRange(id int64, rng extensions.GroupIDRange) bool {
        return id >= rng.Min && id <= rng.Max
 }

-// AllowsHostVolumePath is a utility for checking if a PSP allows the host volume path.
-// This only checks the path. You should still check to make sure the host volume fs type is allowed.
-func AllowsHostVolumePath(psp *extensions.PodSecurityPolicy, hostPath string) bool {
+// AllowsHostVolumePath is a utility for checking if a PSP allows the host volume path, and whether volume mounts
+// that use the path are required to be readonly. This only checks the path. You should still check to make sure the host volume fs type is allowed.
+func AllowsHostVolumePath(psp *extensions.PodSecurityPolicy, hostPath string) (pathIsAllowed, mustBeReadOnly bool) {
        if psp == nil {
-               return false
+               return false, false
        }

        // If no allowed paths are specified then allow any path
        if len(psp.Spec.AllowedHostPaths) == 0 {
-               return true
+               return true, false
        }

        for _, allowedPath := range psp.Spec.AllowedHostPaths {
                if hasPathPrefix(hostPath, allowedPath.PathPrefix) {
-                       return true
+                       if !allowedPath.ReadOnly {
+                               return true, false
+                       }
+                       pathIsAllowed = true
+                       mustBeReadOnly = true
                }
        }

-       return false
+       return pathIsAllowed, mustBeReadOnly
 }
```